JSON Schema OneOf Support

CamBDignan opened the pull request:
**To be merged after https://github.com/CamBDignan/gnostic/pull/3**

**Related issue:** https://github.com/google/gnostic/issues/251

Currently, the `oneof` specifier in proto messages is ignored in the translated JSON schema. For example, the following proto:

```proto
message Pet {
  oneof type_of_pet {
    string cat = 1;
    string dog = 2;
  }
}
```

will be translated to:

```json
{
  "title": "Pet",
  "$id": "Pet.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "cat": {
      "title": "cat",
      "type": "string",
      "default": ""
    },
    "dog": {
      "title": "dog",
      "type": "string",
      "default": ""
    }
  }
}
```

This is undesirable because then the following JSON would be considered valid:

```json
{
  "cat": "Whiskers",
  "dog": "Barky"
}
```

This PR attempts to maintain the spirit of protobuf `oneof` in JSON schema. In order to accomplish this goal, I thought about what we want a valid JSON to actually look like. Continuing with the `Pet` example above, the following could be considered valid:

```json
{
  "cat": "Whiskers"
}
```

or

```json
{
  "dog": "Barky"
}
```

For oneofs in protobuf, it is also possible that neither alternative in the oneof is specified, so maybe in JSON that would just be the following:

```json
{}
```

However, like I mentioned before, we are trying to maintain the spirit of the protobuf `oneof`. When consuming protobuf generated code, we do not check for the presence or absence of a field to determine which alternative is specified. Rather, we do a switch on some enum (in the `Pet` example, we would switch on `type_of_pet`). So maybe JSON that is closer in spirit to the protobuf would look more like the following:

```json
{
  "kind": "cat",
  "value": "Whiskers"
}
```

or

```json
{
  "kind": "dog",
  "value": "Barky"
}
```

(Null, or absence of a value, could still be used to represent "unspecified").

This seems to work pretty well - clients can always switch on the `kind` enum to determine the type of object they are dealing with. However, there is one more thing to consider - a single proto message can contain more than one `oneof`. Consider the following:

```proto
message DateTimeOffset {
  oneof local_date_time {
    sint64 local_epoch_ms = 1;
    DateTimeComponents date_time_components = 2;
  }

  oneof time_offset {
    sint64 utc_offset = 3;
  }
}
```

I realized in order to accommodate the above case, one more level of nesting would be necessary. See the following examples of valid JSON for the above proto for what I mean:

```json
{
  "localDateTime": {
    "kind": "local_epoch_ms",
    "value": 123
  },
  "timeOffset": {
    "kind": "utc_offset",
    "value": 9
  }
}
```

or

```json
{
  "localDateTime": {
    "kind": "date_time_components",
    "value": {
      // ...
    }
  },
  "timeOffset": {
    "kind": "utc_offset",
    "value": 9
  }
}
```

And its easy to see how this additional level of nesting would work for the `Pet` case as well:

```json
{
  "typeOfPet": {
    "kind": "cat",
    "value": "Whiskers"
  }
}
```

or

```json
{
  "typeOfPet": {
    "kind": "dog",
    "value": "Barky"
  }
}  
```

Now that we have an idea what valid JSON would look like, we can attempt to define the JSON schemas.

**Pet.json**

```json
{
  "title": "Pet",
  "$id": "Pet.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "typeOfPet": {
      "oneOf": [
        {
          "type": "null"
        },
        {
          "$ref": "#/definitions/Pet_Cat"
        },
        {
          "$ref": "#/definitions/Pet_Dog"
        }
      ],
      "default": null
    }
  },
  "definitions": {
    "Pet_Cat": {
      "title": "Pet_Cat",
      "type": "object",
      "properties": {
        "kind": {
          "type": "string",
          "enum": [
            "cat"
          ],
          "default": "cat"
        },
        "value": {
          "title": "value",
          "type": "string",
          "default": ""
        }
      }
    },
    "Pet_Dog": {
      "title": "Pet_Dog",
      "type": "object",
      "properties": {
        "kind": {
          "type": "string",
          "enum": [
            "dog"
          ],
          "default": "dog"
        },
        "value": {
          "title": "value",
          "type": "string",
          "default": ""
        }
      }
    }
  }
}
```

**DateTimeOffset.json**

```json
{
  "title": "DateTimeOffset",
  "$id": "DateTimeOffset.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "localDateTime": {
      "oneOf": [
        {
          "type": "null"
        },
        {
          "$ref": "#/definitions/DateTimeOffset_LocalEpochMs"
        },
        {
          "$ref": "#/definitions/DateTimeOffset_DateTimeComponents"
        }
      ],
      "default": null
    },
    "timeOffset": {
      "oneOf": [
        {
          "type": "null"
        },
        {
          "$ref": "#/definitions/DateTimeOffset_UtcOffset"
        }
      ],
      "default": null
    }
  },
  "definitions": {
    "DateTimeOffset_LocalEpochMs": {
      "title": "DateTimeOffset_LocalEpochMs",
      "type": "object",
      "properties": {
        "kind": {
          "type": "string",
          "enum": [
            "local_epoch_ms"
          ],
          "default": "local_epoch_ms"
        },
        "value": {
          "title": "value",
          "type": "integer",
          "default": 0
        }
      }
    },
    "DateTimeOffset_DateTimeComponents": {
      "title": "DateTimeOffset_DateTimeComponents",
      "type": "object",
      "properties": {
        "kind": {
          "type": "string",
          "enum": [
            "date_time_components"
          ],
          "default": "date_time_components"
        },
        "value": {
          "$ref": "DateTimeComponents.json"
        }
      }
    },
    "DateTimeOffset_UtcOffset": {
      "title": "DateTimeOffset_UtcOffset",
      "type": "object",
      "properties": {
        "kind": {
          "type": "string",
          "enum": [
            "utc_offset"
          ],
          "default": "utc_offset"
        },
        "value": {
          "title": "value",
          "type": "integer",
          "default": 0
        }
      }
    }
  }
}
```

CamBDignan committed:
JSON Schema OneOf Support

CamBDignan committed:
new strategy

CamBDignan committed:
wip

CamBDignan committed:
minor fix

CamBDignan committed:
separate out subschemas

CamBDignan committed:
Merge branch 'separate-out-subschemas' into jsonschema-oneof-support

CamBDignan committed:
delete commented code

CamBDignan committed:
Merge branch 'separate-out-subschemas' into jsonschema-oneof-support

CamBDignan committed:
default values

CamBDignan committed:
Merge branch 'default-values' into jsonschema-oneof-support

CamBDignan committed:
improved defaults

CamBDignan committed:
Merge branch 'default-values' into jsonschema-oneof-support

CamBDignan committed:
reorg

CamBDignan committed:
cleanup

CamBDignan committed:
Merge branch 'separate-out-subschemas' into default-values

CamBDignan committed:
Merge branch 'default-values' into jsonschema-oneof-support

CamBDignan committed:
reorg for readability

CamBDignan committed:
fix bug

CamBDignan committed:
Merge branch 'default-values' into jsonschema-oneof-support

CamBDignan committed:
typeOfObject -> kind

CamBDignan commented on cmd/protoc-gen-jsonschema/generator/json-schema.go:
```diff
@@ -266,31 +296,154 @@ func (g *JSONSchemaGenerator) schemaOrReferenceForField(field protoreflect.Field
+
+func (g *JSONSchemaGenerator) buildKindProperty(propertyValue string) *jsonschema.NamedSchema {
+	kindProperty := &jsonschema.NamedSchema{
+		Name: "kind",
```
One point of discussion during our meeting was what to name the discriminator property. I chose `kind` as I have seen examples of it before (although I can't currently find any) and because it is used within Esri. @stephenrhea2 suggested `instanceType` or `subtype` since it is sort of representing a class hierarchy. I have seen examples using `objectType` on the web as well. Any preferences? I personally don't have a strong preference one way or another.

Said-Parirokh replied:
If `kind` is the industry standard, then let's stick with it even if it is kind of ambiguous.

CamBDignan replied:
I don't think it is necessarily the industry standard, but at the same time I don't think that there really is an industry standard for the discriminator name - I have seen quite a few examples and each seems to use something different.

Said-Parirokh replied:
I quite like "discriminator". It captures the intent accurately, and is less likely to collide with users' field names. Though Olivier provided the approach to solve that problem regardless of choice made for the discriminator field name.

stephenrhea2 replied:
I'm still not sold on Olivier's idea of a `kind-value` design because of the JSON serializers that include all properties by default (like C#). Wouldn't the serializer have to be customized to write different properties into the `value` field based on the object's type?

TBH, it feels like I'm missing something since everyone else liked the idea. If that's an invalid concern, feel free to ignore.

CamBDignan replied:
@stephenrhea2 I think it should be fine because even though the name of the field is the same everywhere now (`value`), its really a different field (the only thing in common is the name). JSON serializers should know which properties to write to the field based on which `value` field it actually is (based on the object's type).

If I am misunderstanding your concern then please let me know, an example might be helpful. Thanks!

cc @Olivier11515

CamBDignan replied:
Just to elaborate, let's consider the following example:

```proto
message Identifier {
  oneof type_of_identifier {
    string string_id = 1;
    int64 int_id = 2;
  }
}
```

This maps to the following classes in Python:

```python
class Identifier(BaseModel):
  typeOfIdentifier: Optional[Union[IdentifierStringId, IdentifierIntId]] = None

class IdentifierStringId(BaseModel):
  kind: Literal["type_string_id"] = "string_id"
  value: str = ""

class IdentifierIntId(BaseModel):
  kind: Literal["type_int_id"] = "int_id"
  value: int = 0
```

Although we have two different fields that are both named `value`, they are at the end of the day two different fields in different classes with different types. I don't think JSON serializers should have any issue with this.

CamBDignan replied:
> I quite like "discriminator". It captures the intent accurately, and is less likely to collide with users' field names. Though Olivier provided the approach to solve that problem regardless of choice made for the discriminator field name.

I actually remember now where I saw `kind` being used - https://pkg.go.dev/google.golang.org/protobuf@v1.36.5/reflect/protoreflect#Kind. It is actually used by protobuf client code to distinguish between different kinds of fields.

I think that because `kind` is already used by Esri's AI Assistant Service and because there does seem to be a precedent for it being used elsewhere, I will prefer to just stick with it for now. But I am willing to go with the consensus if others feel differently.

stephenrhea2 replied:
I'll have to think about this more tomorrow. The biggest issue is that C# doesn't have the concept of type unions, but there have got to be other examples of C# using `anyOf` somewhere around the company.

Said-Parirokh replied:
There is not language support for `oneOf` in C#, but the problem has been addressed in 3rd party libraries: 
 - https://github.com/mcintyre321/OneOf

It's also one of the most requested features of C#: 
 - https://github.com/dotnet/csharplang/discussions/8926

I don't think this is enough of a limitation to warrant not moving forward with the proposal. It could be a good idea to attempt to consume the JSON from a C# client to understand how painful it might be.

stephenrhea2 replied:
> It's also one of the most requested features of C#:

A former coworker who preferred TypeScript got really excited when that feature started gaining traction. 😆

Said-Parirokh reviewed (approved):
Looks great! Thanks!

CamBDignan committed:
rename property to value

CamBDignan committed:
inverse conditional

CamBDignan commented:
Regarding the discussion we had today with respect to translating `deprecated` metadata tag from protobuf to JSON schema:

Looks like support for deprecation of properties was not added to JSON schema until `draft 2019-09` - https://stackoverflow.com/questions/48461197/how-do-i-mark-a-property-in-json-schema-as-deprecated. This tool generates an earlier version of JSON schemas - `draft-07`.

I personally don't see it as too big of a deal - it is pretty rarely that we mark fields as deprecated in protobuf anyway. However we can always look into generating a newer version of the JSON schema that supports this feature in the future.

cc @Olivier11515

dheerajv reviewed (approved):
LGTM 👍 🦸  

Hopefully, we'll soon finalize our decision on forking the Gnostic repo and placing it under the WebGIS org or another appropriate org.

CamBDignan committed:
change naming scheme for kind property

CamBDignan committed:
Revert "change naming scheme for kind property"

This reverts commit e950278e1f9c9e795dbe24d5cf8c1a1931285b02.

CamBDignan committed:
replace TypeUnspecified with null

CamBDignan commented:
Summarizing discussion with @Olivier11515, @Said-Parirokh, and @stephenrhea2:

The original proposal was to translate the following proto message:

```proto
message Pet {
  oneof type_of_pet {
    string cat = 1;
    string dog = 2;
  }
}
```

into the following JSON schemas:

**TypeUnspecified.json**

```json
{
  "title": "TypeUnspecified.json",
  "type": "object",
  "properties": {
    "kind": {
      "type": "string",
      "enum": [
        "unspecified"
      ],
      "default": "unspecified"
    }
  }
}
```

**Pet.json**

```json
{
  "title": "Pet.json",
  "type": "object",
  "properties": {
    "typeOfPet": {
      "oneOf": [
        {
          "$ref": "TypeUnspecified.json"
        },
        {
          "$ref": "#/definitions/Pet_Cat"
        },
        {
          "$ref": "#/definitions/Pet_Dog"
        }
      ]
    }
  },
  "definitions": {
    "Pet_Cat": {
      "title": "Pet_Cat",
      "type": "object",
      "properties": {
        "kind": {
          "type": "string",
          "enum": [
            "cat"
          ],
          "default": "cat"
        },
        "value": {
          "type": "string",
          "default": ""
        }
      }
    },
    "Pet_Dog": {
      "title": "Pet_Dog",
      "type": "object",
      "properties": {
        "kind": {
          "type": "string",
          "enum": [
            "dog"
          ],
          "default": "dog"
        },
        "value": {
          "type": "string",
          "default": ""
        }
      }
    }
  }
}
```

@Olivier11515 raised the point that this scheme would not work if the name of the field in the proto oneof is `unspecified`, since we are using `unspecified` as a reserved keyword to mean "no alternative in the oneof is specified". I proposed that we could work around this by prefixing all of the non-unspecified `kind` enum values with something like `type_`, in order to avoid any possible naming collision with `unspecified`. This would work, however it led to a larger discussion about the need for `TypeUnspecified` in the first place.

The main takeaways from the discussion were as follows:

- For oneofs, we need some type to be an acceptable default value if no field is present. We determined this could be `TypeUnspecified` or just `null`. See the possible revised `typeOfPet` properties below to get a sense of what this could look like:

```json
{
  "oneOf": [
    {
      "$ref": "TypeUnspecified.json"
    },
    {
      "$ref": "#/definitions/Pet_Cat"
    },
    {
      "$ref": "#/definitions/Pet_Dog"
    }
  ],
  "default": {
    "kind": "unspecified"
  }
}
```

or

```json
{
  "oneOf": [
    {
      "type": "null"
    },
    {
      "$ref": "#/definitions/Pet_Cat"
    },
    {
      "$ref": "#/definitions/Pet_Dog"
    }
  ],
  "default": null
}
```

- `null` is often used in JSON to indicate the absence of a value. In protobuf, the equivalent concept would be field presence (see https://protobuf.dev/programming-guides/field_presence/). High level summary - every field in protobuf has a non-null default (0 for int, false for bool, "" for string, list with 0 elements for repeated, nested object containing default primitives for message, etc.). There is additionally a flag that can be checked representing whether a value for the field was explicitly set. For oneofs, each alternative also defaults to a non-null value, however rather than a boolean flag, there is an enum indicating which field if any is explicitly set (an enum value of 0 indicating that no field in the oneof was explicitly set).
- For other types besides oneofs, there is a reasonable non-null default in JSON corresponding to the default in protobuf. However, for oneofs the only reasonable default is to assume that no alternative in the oneof is explicitly set.
- There is no need to invent a new type (`TypeUnspecified`) to represent semantically what `null` already represents in JSON. Therefore, we can just use `null` as the default for `oneof` fields and remove `TypeUnspecified`. This also removes the need to the aforementioned `type_` naming scheme for `kind` values.

dheerajv commented:
@CamBDignan 
I am planning to watch the meeting recording today and would like to review this PR.

Olivier11515 commented:
@CamBDignan I read the description and have some clarifying questions and comments wrt "Pet.json/definitions" :

- I wonder if the `kind` properties need a default value? I guess they don't but wanted to confirm my understanding (in the description they have a default value).

- I think all `kind` properties need to be `required`.

- `value` properties that have `"type": "string"` also have `"default": ""` (instead of `"default": null`), is it because `""` is the default string value in proto3 or is it for another reason?
  - Maybe it would be simpler to make the `value` properties `required` as well (like `kind` properties), this way we don't need to define a default value for them, and there is a single way to represent a payload where the value is an empty string?

- the `value` properties have `"title": "value"` but the `kind` properties don't have a `"title"`. I guess we should either have `title` or not have `title` in both cases?

CamBDignan commented:
> * I wonder if the `kind` properties need a default value? I guess they don't but wanted to confirm my understanding (in the description they have a default value).
> * I think all `kind` properties need to be `required`.

These two points are related so I will answer them together. In general, fields need to either be required, or a default behavior needs to be defined (whether or not that default is specified in the JSON schema). In this specific case, it is more appropriate for the `kind` property to not be required, and to instead have a default, because it is only needed in order to distinguish between alternatives in a `oneOf`. In general (when used in isolation), the `kind` property does not need to be specified.

This might become more clear when looking at the generated Python code. Take the following example with a required `kind` field:

```python
class PetCat(BaseModel):
  kind: Literal["cat"]
  value: str = ""
```

and here is the same example with the default for `kind`:

```python
class PetCat(BaseModel):
  kind: Literal["cat"] = "cat"
  value: str = ""
```

The former example unnecessarily requires consuming code to instantiate `PetCat` objects like so:

```python
cat: PetCat = PetCat(
  kind="cat",
  value="Whiskers",
)
```

while with the latter example, this can be simplified as follows:

```python
cat: PetCat = PetCat(
  value="Whiskers",
)
```

> * `value` properties that have `"type": "string"` also have `"default": ""` (instead of `"default": null`), is it because `""` is the default string value in proto3 or is it for another reason?
>   
>   * Maybe it would be simpler to make the `value` properties `required` as well (like `kind` properties), this way we don't need to define a default value for them, and there is a single way to represent a payload where the value is an empty string?

Yes, none of the properties are required, including `value`, for consistency with the behavior in protobuf. Why do you suggest that it would be simpler to make `value` properties `required`? If we made them `required`, we would be making an exception as none of the other properties are `required` anywhere else. Do you have a specific example in mind where having them as `required` makes things simpler?

> * the `value` properties have `"title": "value"` but the `kind` properties don't have a `"title"`. I guess we should either have `title` or not have `title` in both cases?

Yes, I believe that the `kind` property should also have a `title` for consistency with other properties. I will make that change. Thanks!

CamBDignan committed:
title for kind properties

dheerajv commented:
@CamBDignan,  

The design looks great! I had a couple of questions for clarification:

Question 1: How are you planning to handle `optional` primitive types? Will they be treated similarly to `oneof` fields, where the default is `null`?  

Below is an example illustrating my understanding using `optional string`:

```proto
message SampleMessage {
  optional string str_prop = 1;
}
```

Expected JSON schema representation:

```json
{
  "title": "SampleMessage",
  "type": "object",
  "properties": {
    "typeOfStrProp": {
      "oneOf": [
        { "type": "null" },
        { "$ref": "#/definitions/SampleMessage_StrProp" }
      ],
      "default": null
    }
  },
  "definitions": {
    "SampleMessage_StrProp": {
      "title": "SampleMessage_StrProp",
      "type": "object",
      "properties": {
        "kind": {
          "type": "string",
          "enum": ["str_prop"],
          "default": "str_prop"
        },
        "value": {
          "type": "string",
          "default": ""
        }
      }
    }
  }
}
```
Is this aligned with your current approach?


Question 2: In the case where a message contains another message, like the example below, clients may choose not to set the inner message. In generated code from proto, we can check this using `hasInnerMsg()`.

How would this translate when using Pydantic classes?

Example:
```proto
message InnerMessage {
  string str_prop = 1;
}

message OuterMessage {
  InnerMessage inner_msg = 1;
  bool bool_prop = 2;
}
```

Olivier11515 commented:
> > * `value` properties that have `"type": "string"` also have `"default": ""` (instead of `"default": null`), is it because `""` is the default string value in proto3 or is it for another reason?
> >   
> >   * Maybe it would be simpler to make the `value` properties `required` as well (like `kind` properties), this way we don't need to define a default value for them, and there is a single way to represent a payload where the value is an empty string?
> 
> Yes, none of the properties are required, including `value`, for consistency with the behavior in protobuf.

I don't think making `kind` and `value` properties `optional` replicates the protobuf behaviour. There is probably a misunderstanding on what `required` means in json-schema.

`required` in json schema means "if the owning object is there, the property must be there too". It doesn't mean "the property must be there all the time".

For example, if we make `value` required, then this is valid (and corresponds to the case where the oneof in protobuf is a `dog`):

```json
{
  "typeOfPet": {
    "kind": "dog",
    "value": "Barky"
  }
} 
```

this is valid (and corresponds to the case where the oneof in protobuf is not set):

```json
{
} 
```

this is not valid (and corresponds to nothing in protobuf):

```json
{
  "typeOfPet": {
    "kind": "dog",
  }
} 
```

So making `value` a `required` property makes things consistent with the oneof behaviour in protobuf.

> Why do you suggest that it would be simpler to make `value` properties `required`?

It is simpler because it eliminates a case that corresponds to nothing in protobuf (the last one above).

> If we made them `required`, we would be making an exception as none of the other properties are `required` anywhere else.

The other properties are `optional` because they correspond to protobuf fields that may or may not be set.

The `oneof` may or may not be set as well, and this is represented by `typeOfPet` being `optional`.

`value` (and `kind`) are one level deeper than other properties, hence they don't follow the same rules. They are owned by `typeOfPet`.

> Do you have a specific example in mind where having them as `required` makes things simpler?

See the example above.

Olivier11515 commented:
@CamBDignan Another reason why I like the idea of having `kind` required is that this JSON:
```json
{
  "typeOfPet": {
  }
} 
```
is currently ambiguous I think as per the json-schema from the description: it could either be a `cat` or a `dog`.

I think this one is ambiguous as well:

```json
{
  "typeOfPet": {
    "value" : "whoami"
  }
} 
```

Note: I saw the explanation you wrote in https://github.com/CamBDignan/gnostic/pull/1#issuecomment-2752441681 showing that having a default value for `kind` results in better code generation. But I guess you could implement a code generation rule to benefit from the same behaviour when a property is `required` and defined as an `enum` with a single value in it: in this case the corresponding python field would be initialized to this single value because no valid JSON can contain any other value for that field.

Side note: why are we using an [`enum`](https://json-schema.org/understanding-json-schema/reference/enum) with a single value instead of [`const`](https://json-schema.org/understanding-json-schema/reference/const) ?

CamBDignan commented:
> Question 1: How are you planning to handle `optional` primitive types? Will they be treated similarly to `oneof` fields, where the default is `null`?

Yes, your example illustrates how we will handle it, it is the same as for oneofs.

> Question 2: In the case where a message contains another message, like the example below, clients may choose not to set the inner message. In generated code from proto, we can check this using `hasInnerMsg()`.
> 
> How would this translate when using Pydantic classes?

It would be something like this:

```python
class InnerMessage(BaseModel):
  strProp: str = ""

class OuterMessage(BaseModel):
  innerMsg: Optional[InnerMessage] = None
  boolProp: bool = False
```

`innerMsg` being `None` is equivalent to `hasInnerMsg() == false`

CamBDignan commented:
> this is not valid (and corresponds to nothing in protobuf):
> 
> ```json
> {
>   "typeOfPet": {
>     "kind": "dog",
>   }
> } 
> ```
>
> So making `value` a `required` property makes things consistent with the oneof behaviour in protobuf.

I understand that we technically can make value required, and therefore this case would be invalid, but I don't understand why we would want to. Yes, I understand that typically if the `typeOfPet` object is present, it would be because you are actually specifying a value. However, I don't see any reason to force clients to do so. And I would argue it is more consistent with the protobuf behavior to have defaults for fields when they are not present rather than make them required, even in this case.

> > Why do you suggest that it would be simpler to make `value` properties `required`?
> 
> It is simpler because it eliminates a case that corresponds to nothing in protobuf (the last one above).

Please see my previous point. I would be open to making the field required if there was a really strong reason to do so, but in my opinion this is not a strong enough reason.

CamBDignan commented:
> @CamBDignan Another reason why I like the idea of having `kind` required is that this JSON:
> 
> ```json
> {
>   "typeOfPet": {
>   }
> } 
> ```
> 
> is currently ambiguous I think as per the json-schema from the description: it could either be a `cat` or a `dog`.
> 
> I think this one is ambiguous as well:
> 
> ```json
> {
>   "typeOfPet": {
>     "value" : "whoami"
>   }
> } 
> ```

Yes, these are ambiguous in the context of a `oneOf` and therefore would be rejected by the server. `oneOf` means "exactly one of the schemas should match". Therefore, if two or more schemas match it is invalid. So in the context of `oneOf`, the `kind` field must be specified in order to resolve the ambiguity.

In other words, we don't need to explicitly make the `kind` field `required`, but it must be present when the schema is used within a `oneOf` to resolve ambiguities. The `kind` field here is acting as a discriminator field - it is a common practice and the way I have implemented it is how I have seen it implemented in other places.

> Side note: why are we using an [`enum`](https://json-schema.org/understanding-json-schema/reference/enum) with a single value instead of [`const`](https://json-schema.org/understanding-json-schema/reference/const) ?

I originally had it both as an enum with a single value and with `const` - this is consistent with how I have seen FastAPI generate `oneOfs` with a discriminator in their OpenAPI spec. However, I removed `const` because it was causing issues with Python code generation (I don't remember the exact issue, it was a while ago). I don't think it is needed since we have the enum with a single value anyway, but I can always add it back if the code generation issues are resolved in the future, that should not be a breaking change.

CamBDignan commented:
@dheerajv @Olivier11515 I believe I have addressed all of your comments. Please take another look when you get the chance, thanks!

dheerajv reviewed (approved):
LGTM 🎖️

Olivier11515 commented:
Adding meaningful constraints in json-schema makes it easier for clients that write the JSON to understand the semantic.

But I wont keep this PR from being merged. I believe it has been approved so I don't need to give an approval.

CamBDignan commented:
Thanks everyone for the invaluable feedback on this PR! I believe that the generated JSON schemas have been substantially improved as a result of our discussions!

CamBDignan committed:
Merge branch 'main' into jsonschema-oneof-support
